src/Microsoft.Health.Dicom.Functions/Indexing/ReindexOperation.cs

```diff
@@ -0,0 +1,141 @@
+            EnsureArg.IsNotNull(addExtendedQueryTagService, nameof(addExtendedQueryTagService));
+            _reindexConfig = configOptions.Value.Add;
+            _tagOperationStore = tagOperationStore;
+            _instanceReindexer = instanceReindexer;
+            _addExtendedQueryTagService = addExtendedQueryTagService;
+        }
+
+        /// <summary>
+        ///  The orchestration function to add extended query tags.
+        /// </summary>
+        /// <param name="context">The context.</param>
+        /// <param name="logger">The logger.</param>
+        /// <returns>The task.</returns>
+        [FunctionName(nameof(AddExtendedQueryTagsOrchestrationAsync))]
+        public async Task AddExtendedQueryTagsOrchestrationAsync(
+            [OrchestrationTrigger] IDurableOrchestrationContext context,
+            ILogger logger)
+        {
+            EnsureArg.IsNotNull(context, nameof(context));
+            logger = context.CreateReplaySafeLogger(logger);
+            var input = context.GetInput<IEnumerable<AddExtendedQueryTagEntry>>();
+            var tagEntries = await context.CallActivityAsync<IEnumerable<ExtendedQueryTagStoreEntry>>(nameof(AddExtendedQueryTagAsync), input);
+            await context.CallSubOrchestratorAsync(nameof(ReindexExtendedQueryTagsOrchestrationAsync), context.InstanceId, tagEntries);
+        }
+
+        /// <summary>
+        /// The orchestration function to Reindex Extended Query Tags.
+        /// </summary>
+        /// <param name="context">the context.</param>
+        /// <param name="logger">The logger.</param>
+        /// <returns></returns>
+        [FunctionName(nameof(ReindexExtendedQueryTagsOrchestrationAsync))]
+        public async Task ReindexExtendedQueryTagsOrchestrationAsync(
+           [OrchestrationTrigger] IDurableOrchestrationContext context,
+           ILogger logger)
+        {
+            EnsureArg.IsNotNull(context, nameof(context));
+            logger = context.CreateReplaySafeLogger(logger);
+            var input = context.GetInput<IEnumerable<ExtendedQueryTagStoreEntry>>();
+
+
+            // TODO: should start a new orchestraion instead of while loop
+            while (true)
+            {
+                IReadOnlyList<long> watermarks = await context
+                    .CallActivityAsync<IReadOnlyList<long>>(nameof(GetNextWatermarksOfOperationAsync), context.InstanceId);
+                if (watermarks.Count == 0)
+                {
+                    break;
+                }
+
+                IReadOnlyList<ExtendedQueryTagStoreEntry> queryTags = await context
+                    .CallActivityAsync<IReadOnlyList<ExtendedQueryTagStoreEntry>>(nameof(GetExtendedQueryTagsOfOperationAsync), context.InstanceId);
+
+                if (queryTags.Count == 0)
+                {
+                    break;
+                }
+
+                // Reindex
+                // TODO: process them parallel
```

Comment:
Is there a particular reason we're not doing this now? Just for testing?

Reply:
It's not hard to make it parallel now, just want to build the min system which could up and running at this moment.

I'm not worry about this part, once the system start working, I'll add the parallel.